Coordinated multiple views

keller-mark opened the pull request:
Demo: https://s3.amazonaws.com/vitessce-data/demos/2020-08-27/a80a7a5/index.html
Overview: https://github.com/hubmapconsortium/vitessce/blob/next/NEXT.md

The Linnarsson and Wang demos have been "translated" to the new view config schema in api.js, while the other view configs keep the "legacy" schema (with the addition of a `version` property) which is made possible by an `upgrade()` function.

Fixes #465
Fixes #574 
Fixes #733 
Towards #30 
Towards #189
Towards #716
Towards #724 
Towards #725 
Replaces #717 
Replaces #727 

No rush to get this reviewed, I realize this pull request is very large. Happy to discuss it on zoom as well.


More demos (corresponding to the use cases outlined in the google doc linked from https://github.com/hubmapconsortium/vitessce/issues/465):
- [Linnarsson with two spatial views, linked on only zoom and x-axis pan](https://s3.amazonaws.com/vitessce-data/demos/2020-08-27/a80a7a5/index.html?url=data%3A%2C%7B%22name%22%3A%22Linnarsson%22%2C%22version%22%3A%221.0.0%22%2C%22description%22%3A%22Spatial%20organization%20of%20the%20somatosensory%20cortex%20revealed%20by%20cyclic%20smFISH%22%2C%22public%22%3Atrue%2C%22datasets%22%3A%5B%7B%22uid%22%3A%22linnarsson-2018%22%2C%22name%22%3A%22Linnarsson%202018%22%2C%22files%22%3A%5B%7B%22type%22%3A%22cells%22%2C%22fileType%22%3A%22cells.json%22%2C%22url%22%3A%22https%3A%2F%2Fs3.amazonaws.com%2Fvitessce-data%2F0.0.31%2Fmaster_release%2Flinnarsson%2Flinnarsson.cells.json%22%7D%2C%7B%22type%22%3A%22cell-sets%22%2C%22fileType%22%3A%22cell-sets.json%22%2C%22url%22%3A%22https%3A%2F%2Fs3.amazonaws.com%2Fvitessce-data%2F0.0.31%2Fmaster_release%2Flinnarsson%2Flinnarsson.cell-sets.json%22%7D%2C%7B%22type%22%3A%22raster%22%2C%22fileType%22%3A%22raster.json%22%2C%22url%22%3A%22https%3A%2F%2Fs3.amazonaws.com%2Fvitessce-data%2F0.0.31%2Fmaster_release%2Flinnarsson%2Flinnarsson.raster.json%22%7D%2C%7B%22type%22%3A%22molecules%22%2C%22fileType%22%3A%22molecules.json%22%2C%22url%22%3A%22https%3A%2F%2Fs3.amazonaws.com%2Fvitessce-data%2F0.0.31%2Fmaster_release%2Flinnarsson%2Flinnarsson.molecules.json%22%7D%2C%7B%22type%22%3A%22neighborhoods%22%2C%22fileType%22%3A%22neighborhoods.json%22%2C%22url%22%3A%22https%3A%2F%2Fs3.amazonaws.com%2Fvitessce-data%2F0.0.31%2Fmaster_release%2Flinnarsson%2Flinnarsson.neighborhoods.json%22%7D%2C%7B%22type%22%3A%22expression-matrix%22%2C%22fileType%22%3A%22clusters.json%22%2C%22url%22%3A%22https%3A%2F%2Fs3.amazonaws.com%2Fvitessce-data%2F0.0.31%2Fmaster_release%2Flinnarsson%2Flinnarsson.clusters.json%22%7D%5D%7D%5D%2C%22initStrategy%22%3A%22auto%22%2C%22coordinationSpace%22%3A%7B%22embeddingZoom%22%3A%7B%22PCA%22%3A0%2C%22TSNE%22%3A0.75%7D%2C%22embeddingType%22%3A%7B%22PCA%22%3A%22PCA%22%2C%22TSNE%22%3A%22t-SNE%22%7D%2C%22spatialZoom%22%3A%7B%22A%22%3A-5.5%7D%2C%22spatialTargetX%22%3A%7B%22A%22%3A16000%7D%2C%22spatialTargetY%22%3A%7B%22A%22%3A20000%2C%22B%22%3A20000%7D%2C%22spatialLayers%22%3A%7B%22B%22%3A%5B%7B%22type%22%3A%22cells%22%2C%22stroked%22%3Afalse%2C%22visible%22%3Atrue%2C%22radius%22%3A50%7D%5D%7D%7D%2C%22layout%22%3A%5B%7B%22component%22%3A%22description%22%2C%22props%22%3A%7B%22description%22%3A%22Linnarsson%3A%20Spatial%20organization%20of%20the%20somatosensory%20cortex%20revealed%20by%20cyclic%20smFISH%22%7D%2C%22x%22%3A0%2C%22y%22%3A0%2C%22w%22%3A2%2C%22h%22%3A1%7D%2C%7B%22component%22%3A%22layerController%22%2C%22x%22%3A0%2C%22y%22%3A1%2C%22w%22%3A2%2C%22h%22%3A4%2C%22coordinationScopes%22%3A%7B%22spatialLayers%22%3A%22A%22%7D%7D%2C%7B%22component%22%3A%22status%22%2C%22x%22%3A0%2C%22y%22%3A5%2C%22w%22%3A2%2C%22h%22%3A1%7D%2C%7B%22component%22%3A%22spatial%22%2C%22coordinationScopes%22%3A%7B%22spatialZoom%22%3A%22A%22%2C%22spatialTargetX%22%3A%22A%22%2C%22spatialTargetY%22%3A%22A%22%2C%22spatialLayers%22%3A%22A%22%7D%2C%22x%22%3A2%2C%22y%22%3A0%2C%22w%22%3A4%2C%22h%22%3A2%7D%2C%7B%22component%22%3A%22spatial%22%2C%22coordinationScopes%22%3A%7B%22spatialZoom%22%3A%22A%22%2C%22spatialTargetX%22%3A%22A%22%2C%22spatialTargetY%22%3A%22B%22%2C%22spatialLayers%22%3A%22B%22%7D%2C%22x%22%3A2%2C%22y%22%3A2%2C%22w%22%3A4%2C%22h%22%3A2%7D%2C%7B%22component%22%3A%22genes%22%2C%22x%22%3A9%2C%22y%22%3A0%2C%22w%22%3A3%2C%22h%22%3A2%7D%2C%7B%22component%22%3A%22cellSets%22%2C%22x%22%3A9%2C%22y%22%3A3%2C%22w%22%3A3%2C%22h%22%3A2%7D%2C%7B%22component%22%3A%22heatmap%22%2C%22props%22%3A%7B%22transpose%22%3Atrue%7D%2C%22x%22%3A2%2C%22y%22%3A4%2C%22w%22%3A10%2C%22h%22%3A2%7D%2C%7B%22component%22%3A%22scatterplot%22%2C%22coordinationScopes%22%3A%7B%22embeddingType%22%3A%22PCA%22%2C%22embeddingZoom%22%3A%22PCA%22%7D%2C%22x%22%3A6%2C%22y%22%3A0%2C%22w%22%3A3%2C%22h%22%3A2%7D%2C%7B%22component%22%3A%22scatterplot%22%2C%22coordinationScopes%22%3A%7B%22embeddingType%22%3A%22TSNE%22%2C%22embeddingZoom%22%3A%22TSNE%22%7D%2C%22x%22%3A6%2C%22y%22%3A2%2C%22w%22%3A3%2C%22h%22%3A2%7D%5D%7D)
    - spatial layer controller is only linked to the top spatial view
    - spatial layers for bottom spatial view are pre-defined in the view config (only cell segmentations layer)
- [Cell-set-selection-linked scatterplots and spatial views, gene-selection-linked scatterplots and spatial views](https://s3.amazonaws.com/vitessce-data/demos/2020-08-27/a80a7a5/index.html?url=data%3A%2C%7B%22name%22%3A%22Linnarsson%22%2C%22version%22%3A%221.0.0%22%2C%22description%22%3A%22Spatial%20organization%20of%20the%20somatosensory%20cortex%20revealed%20by%20cyclic%20smFISH%22%2C%22public%22%3Atrue%2C%22datasets%22%3A%5B%7B%22uid%22%3A%22linnarsson-2018%22%2C%22name%22%3A%22Linnarsson%202018%22%2C%22files%22%3A%5B%7B%22type%22%3A%22cells%22%2C%22fileType%22%3A%22cells.json%22%2C%22url%22%3A%22https%3A%2F%2Fs3.amazonaws.com%2Fvitessce-data%2F0.0.31%2Fmaster_release%2Flinnarsson%2Flinnarsson.cells.json%22%7D%2C%7B%22type%22%3A%22cell-sets%22%2C%22fileType%22%3A%22cell-sets.json%22%2C%22url%22%3A%22https%3A%2F%2Fs3.amazonaws.com%2Fvitessce-data%2F0.0.31%2Fmaster_release%2Flinnarsson%2Flinnarsson.cell-sets.json%22%7D%2C%7B%22type%22%3A%22raster%22%2C%22fileType%22%3A%22raster.json%22%2C%22url%22%3A%22https%3A%2F%2Fs3.amazonaws.com%2Fvitessce-data%2F0.0.31%2Fmaster_release%2Flinnarsson%2Flinnarsson.raster.json%22%7D%2C%7B%22type%22%3A%22molecules%22%2C%22fileType%22%3A%22molecules.json%22%2C%22url%22%3A%22https%3A%2F%2Fs3.amazonaws.com%2Fvitessce-data%2F0.0.31%2Fmaster_release%2Flinnarsson%2Flinnarsson.molecules.json%22%7D%2C%7B%22type%22%3A%22neighborhoods%22%2C%22fileType%22%3A%22neighborhoods.json%22%2C%22url%22%3A%22https%3A%2F%2Fs3.amazonaws.com%2Fvitessce-data%2F0.0.31%2Fmaster_release%2Flinnarsson%2Flinnarsson.neighborhoods.json%22%7D%2C%7B%22type%22%3A%22expression-matrix%22%2C%22fileType%22%3A%22clusters.json%22%2C%22url%22%3A%22https%3A%2F%2Fs3.amazonaws.com%2Fvitessce-data%2F0.0.31%2Fmaster_release%2Flinnarsson%2Flinnarsson.clusters.json%22%7D%5D%7D%5D%2C%22initStrategy%22%3A%22auto%22%2C%22coordinationSpace%22%3A%7B%22embeddingZoom%22%3A%7B%22TSNE%22%3A0.75%2C%22PCA%22%3A0%7D%2C%22embeddingType%22%3A%7B%22TSNE%22%3A%22t-SNE%22%2C%22PCA%22%3A%22PCA%22%7D%2C%22embeddingTargetX%22%3A%7B%22TSNE%22%3A0%2C%22PCA%22%3A0%7D%2C%22embeddingTargetY%22%3A%7B%22TSNE%22%3A0%2C%22PCA%22%3A0%7D%2C%22spatialZoom%22%3A%7B%22A%22%3A-5.5%7D%2C%22spatialTargetX%22%3A%7B%22A%22%3A16000%7D%2C%22spatialTargetY%22%3A%7B%22A%22%3A20000%7D%2C%22spatialLayers%22%3A%7B%22A%22%3A%5B%7B%22type%22%3A%22cells%22%2C%22radius%22%3A50%2C%22visible%22%3Atrue%2C%22opacity%22%3A1%2C%22stroked%22%3Afalse%7D%5D%7D%2C%22cellColorEncoding%22%3A%7B%22A%22%3A%22cellSetSelection%22%2C%22B%22%3A%22geneSelection%22%7D%2C%22geneSelection%22%3A%7B%22B%22%3A%5B%22Sox10%22%5D%7D%7D%2C%22layout%22%3A%5B%7B%22component%22%3A%22description%22%2C%22props%22%3A%7B%22description%22%3A%22Linnarsson%3A%20Spatial%20organization%20of%20the%20somatosensory%20cortex%20revealed%20by%20cyclic%20smFISH%22%7D%2C%22x%22%3A0%2C%22y%22%3A8%2C%22w%22%3A3%2C%22h%22%3A4%7D%2C%7B%22component%22%3A%22status%22%2C%22x%22%3A9%2C%22y%22%3A8%2C%22w%22%3A3%2C%22h%22%3A4%7D%2C%7B%22component%22%3A%22genes%22%2C%22coordinationScopes%22%3A%7B%22cellColorEncoding%22%3A%22B%22%2C%22geneSelection%22%3A%22B%22%7D%2C%22x%22%3A0%2C%22y%22%3A0%2C%22w%22%3A3%2C%22h%22%3A8%7D%2C%7B%22component%22%3A%22spatial%22%2C%22coordinationScopes%22%3A%7B%22spatialZoom%22%3A%22A%22%2C%22spatialTargetX%22%3A%22A%22%2C%22spatialTargetY%22%3A%22A%22%2C%22spatialLayers%22%3A%22A%22%2C%22cellColorEncoding%22%3A%22A%22%7D%2C%22x%22%3A6%2C%22y%22%3A8%2C%22w%22%3A3%2C%22h%22%3A4%7D%2C%7B%22component%22%3A%22cellSets%22%2C%22coordinationScopes%22%3A%7B%22cellColorEncoding%22%3A%22A%22%7D%2C%22x%22%3A9%2C%22y%22%3A0%2C%22w%22%3A3%2C%22h%22%3A8%7D%2C%7B%22component%22%3A%22scatterplot%22%2C%22coordinationScopes%22%3A%7B%22embeddingType%22%3A%22TSNE%22%2C%22embeddingZoom%22%3A%22TSNE%22%2C%22embeddingTargetX%22%3A%22TSNE%22%2C%22embeddingTargetY%22%3A%22TSNE%22%2C%22cellColorEncoding%22%3A%22A%22%7D%2C%22x%22%3A6%2C%22y%22%3A0%2C%22w%22%3A3%2C%22h%22%3A4%7D%2C%7B%22component%22%3A%22scatterplot%22%2C%22coordinationScopes%22%3A%7B%22embeddingType%22%3A%22PCA%22%2C%22embeddingZoom%22%3A%22PCA%22%2C%22embeddingTargetX%22%3A%22PCA%22%2C%22embeddingTargetY%22%3A%22PCA%22%2C%22cellColorEncoding%22%3A%22A%22%7D%2C%22x%22%3A6%2C%22y%22%3A4%2C%22w%22%3A3%2C%22h%22%3A4%7D%2C%7B%22component%22%3A%22scatterplot%22%2C%22coordinationScopes%22%3A%7B%22embeddingType%22%3A%22TSNE%22%2C%22embeddingZoom%22%3A%22TSNE%22%2C%22embeddingTargetX%22%3A%22TSNE%22%2C%22embeddingTargetY%22%3A%22TSNE%22%2C%22cellColorEncoding%22%3A%22B%22%2C%22geneSelection%22%3A%22B%22%7D%2C%22x%22%3A3%2C%22y%22%3A0%2C%22w%22%3A3%2C%22h%22%3A4%7D%2C%7B%22component%22%3A%22scatterplot%22%2C%22coordinationScopes%22%3A%7B%22embeddingType%22%3A%22PCA%22%2C%22embeddingZoom%22%3A%22PCA%22%2C%22embeddingTargetX%22%3A%22PCA%22%2C%22embeddingTargetY%22%3A%22PCA%22%2C%22cellColorEncoding%22%3A%22B%22%2C%22geneSelection%22%3A%22B%22%7D%2C%22x%22%3A3%2C%22y%22%3A4%2C%22w%22%3A3%2C%22h%22%3A4%7D%2C%7B%22component%22%3A%22spatial%22%2C%22coordinationScopes%22%3A%7B%22spatialZoom%22%3A%22A%22%2C%22spatialTargetX%22%3A%22A%22%2C%22spatialTargetY%22%3A%22A%22%2C%22spatialLayers%22%3A%22A%22%2C%22cellColorEncoding%22%3A%22B%22%2C%22geneSelection%22%3A%22B%22%7D%2C%22x%22%3A3%2C%22y%22%3A8%2C%22w%22%3A3%2C%22h%22%3A4%7D%5D%7D)
- [Multiple unlinked views of the same image](https://s3.amazonaws.com/vitessce-data/demos/2020-08-27/a80a7a5/index.html?url=data%3A%2C%7B%22name%22%3A%22Linnarsson%22%2C%22version%22%3A%221.0.0%22%2C%22description%22%3A%22Spatial%20organization%20of%20the%20somatosensory%20cortex%20revealed%20by%20cyclic%20smFISH%22%2C%22public%22%3Atrue%2C%22datasets%22%3A%5B%7B%22uid%22%3A%22linnarsson-2018%22%2C%22name%22%3A%22Linnarsson%202018%22%2C%22files%22%3A%5B%7B%22type%22%3A%22cells%22%2C%22fileType%22%3A%22cells.json%22%2C%22url%22%3A%22https%3A%2F%2Fs3.amazonaws.com%2Fvitessce-data%2F0.0.31%2Fmaster_release%2Flinnarsson%2Flinnarsson.cells.json%22%7D%2C%7B%22type%22%3A%22cell-sets%22%2C%22fileType%22%3A%22cell-sets.json%22%2C%22url%22%3A%22https%3A%2F%2Fs3.amazonaws.com%2Fvitessce-data%2F0.0.31%2Fmaster_release%2Flinnarsson%2Flinnarsson.cell-sets.json%22%7D%2C%7B%22type%22%3A%22raster%22%2C%22fileType%22%3A%22raster.json%22%2C%22url%22%3A%22https%3A%2F%2Fs3.amazonaws.com%2Fvitessce-data%2F0.0.31%2Fmaster_release%2Flinnarsson%2Flinnarsson.raster.json%22%7D%2C%7B%22type%22%3A%22molecules%22%2C%22fileType%22%3A%22molecules.json%22%2C%22url%22%3A%22https%3A%2F%2Fs3.amazonaws.com%2Fvitessce-data%2F0.0.31%2Fmaster_release%2Flinnarsson%2Flinnarsson.molecules.json%22%7D%2C%7B%22type%22%3A%22neighborhoods%22%2C%22fileType%22%3A%22neighborhoods.json%22%2C%22url%22%3A%22https%3A%2F%2Fs3.amazonaws.com%2Fvitessce-data%2F0.0.31%2Fmaster_release%2Flinnarsson%2Flinnarsson.neighborhoods.json%22%7D%2C%7B%22type%22%3A%22expression-matrix%22%2C%22fileType%22%3A%22clusters.json%22%2C%22url%22%3A%22https%3A%2F%2Fs3.amazonaws.com%2Fvitessce-data%2F0.0.31%2Fmaster_release%2Flinnarsson%2Flinnarsson.clusters.json%22%7D%5D%7D%5D%2C%22initStrategy%22%3A%22auto%22%2C%22coordinationSpace%22%3A%7B%22spatialZoom%22%3A%7B%22A%22%3A-5.5%2C%22B%22%3A-5.5%7D%2C%22spatialTargetX%22%3A%7B%22A%22%3A16000%2C%22B%22%3A16000%7D%2C%22spatialTargetY%22%3A%7B%22A%22%3A20000%2C%22B%22%3A20000%7D%2C%22spatialLayers%22%3A%7B%22A%22%3A%5B%7B%22type%22%3A%22raster%22%2C%22index%22%3A0%2C%22colormap%22%3A%22%22%2C%22opacity%22%3A1%2C%22visible%22%3Atrue%2C%22channels%22%3A%5B%7B%22selection%22%3A%7B%22channel%22%3A0%7D%2C%22color%22%3A%5B0%2C255%2C0%5D%7D%5D%7D%5D%2C%22B%22%3A%5B%7B%22type%22%3A%22raster%22%2C%22index%22%3A0%2C%22colormap%22%3A%22%22%2C%22opacity%22%3A1%2C%22visible%22%3Atrue%2C%22channels%22%3A%5B%7B%22selection%22%3A%7B%22channel%22%3A0%7D%2C%22color%22%3A%5B0%2C255%2C0%5D%7D%5D%7D%5D%7D%7D%2C%22layout%22%3A%5B%7B%22component%22%3A%22description%22%2C%22props%22%3A%7B%22description%22%3A%22Linnarsson%3A%20Spatial%20organization%20of%20the%20somatosensory%20cortex%20revealed%20by%20cyclic%20smFISH%22%7D%2C%22x%22%3A0%2C%22y%22%3A8%2C%22w%22%3A2%2C%22h%22%3A4%7D%2C%7B%22component%22%3A%22layerController%22%2C%22coordinationScopes%22%3A%7B%22spatialLayers%22%3A%22A%22%7D%2C%22x%22%3A0%2C%22y%22%3A0%2C%22w%22%3A2%2C%22h%22%3A8%7D%2C%7B%22component%22%3A%22layerController%22%2C%22coordinationScopes%22%3A%7B%22spatialLayers%22%3A%22B%22%7D%2C%22x%22%3A10%2C%22y%22%3A0%2C%22w%22%3A2%2C%22h%22%3A8%7D%2C%7B%22component%22%3A%22status%22%2C%22x%22%3A10%2C%22y%22%3A8%2C%22w%22%3A2%2C%22h%22%3A4%7D%2C%7B%22component%22%3A%22spatial%22%2C%22coordinationScopes%22%3A%7B%22spatialZoom%22%3A%22A%22%2C%22spatialTargetX%22%3A%22A%22%2C%22spatialTargetY%22%3A%22A%22%2C%22spatialLayers%22%3A%22A%22%7D%2C%22x%22%3A2%2C%22y%22%3A0%2C%22w%22%3A4%2C%22h%22%3A12%7D%2C%7B%22component%22%3A%22spatial%22%2C%22coordinationScopes%22%3A%7B%22spatialZoom%22%3A%22B%22%2C%22spatialTargetX%22%3A%22B%22%2C%22spatialTargetY%22%3A%22B%22%2C%22spatialLayers%22%3A%22B%22%7D%2C%22x%22%3A6%2C%22y%22%3A0%2C%22w%22%3A4%2C%22h%22%3A12%7D%5D%7D)
- [Pan-linked views of single image](https://s3.amazonaws.com/vitessce-data/demos/2020-08-27/a80a7a5/index.html?url=data%3A%2C%7B%22name%22%3A%22Linnarsson%22%2C%22version%22%3A%221.0.0%22%2C%22description%22%3A%22Spatial%20organization%20of%20the%20somatosensory%20cortex%20revealed%20by%20cyclic%20smFISH%22%2C%22public%22%3Atrue%2C%22datasets%22%3A%5B%7B%22uid%22%3A%22linnarsson-2018%22%2C%22name%22%3A%22Linnarsson%202018%22%2C%22files%22%3A%5B%7B%22type%22%3A%22cells%22%2C%22fileType%22%3A%22cells.json%22%2C%22url%22%3A%22https%3A%2F%2Fs3.amazonaws.com%2Fvitessce-data%2F0.0.31%2Fmaster_release%2Flinnarsson%2Flinnarsson.cells.json%22%7D%2C%7B%22type%22%3A%22cell-sets%22%2C%22fileType%22%3A%22cell-sets.json%22%2C%22url%22%3A%22https%3A%2F%2Fs3.amazonaws.com%2Fvitessce-data%2F0.0.31%2Fmaster_release%2Flinnarsson%2Flinnarsson.cell-sets.json%22%7D%2C%7B%22type%22%3A%22raster%22%2C%22fileType%22%3A%22raster.json%22%2C%22url%22%3A%22https%3A%2F%2Fs3.amazonaws.com%2Fvitessce-data%2F0.0.31%2Fmaster_release%2Flinnarsson%2Flinnarsson.raster.json%22%7D%2C%7B%22type%22%3A%22molecules%22%2C%22fileType%22%3A%22molecules.json%22%2C%22url%22%3A%22https%3A%2F%2Fs3.amazonaws.com%2Fvitessce-data%2F0.0.31%2Fmaster_release%2Flinnarsson%2Flinnarsson.molecules.json%22%7D%2C%7B%22type%22%3A%22neighborhoods%22%2C%22fileType%22%3A%22neighborhoods.json%22%2C%22url%22%3A%22https%3A%2F%2Fs3.amazonaws.com%2Fvitessce-data%2F0.0.31%2Fmaster_release%2Flinnarsson%2Flinnarsson.neighborhoods.json%22%7D%2C%7B%22type%22%3A%22expression-matrix%22%2C%22fileType%22%3A%22clusters.json%22%2C%22url%22%3A%22https%3A%2F%2Fs3.amazonaws.com%2Fvitessce-data%2F0.0.31%2Fmaster_release%2Flinnarsson%2Flinnarsson.clusters.json%22%7D%5D%7D%5D%2C%22initStrategy%22%3A%22auto%22%2C%22coordinationSpace%22%3A%7B%22spatialZoom%22%3A%7B%22A%22%3A-5.5%2C%22B%22%3A-5.5%7D%2C%22spatialTargetX%22%3A%7B%22A%22%3A16000%7D%2C%22spatialTargetY%22%3A%7B%22A%22%3A20000%7D%2C%22spatialLayers%22%3A%7B%22A%22%3A%5B%7B%22type%22%3A%22raster%22%2C%22index%22%3A0%2C%22visible%22%3Atrue%2C%22opacity%22%3A1%2C%22colormap%22%3A%22%22%2C%22channels%22%3A%5B%7B%22selection%22%3A%7B%22channel%22%3A0%7D%2C%22color%22%3A%5B0%2C255%2C0%5D%7D%5D%7D%5D%7D%7D%2C%22layout%22%3A%5B%7B%22component%22%3A%22description%22%2C%22props%22%3A%7B%22description%22%3A%22Linnarsson%3A%20Spatial%20organization%20of%20the%20somatosensory%20cortex%20revealed%20by%20cyclic%20smFISH%22%7D%2C%22x%22%3A0%2C%22y%22%3A0%2C%22w%22%3A2%2C%22h%22%3A1%7D%2C%7B%22component%22%3A%22layerController%22%2C%22x%22%3A0%2C%22y%22%3A1%2C%22w%22%3A2%2C%22h%22%3A4%2C%22coordinationScopes%22%3A%7B%22spatialLayers%22%3A%22A%22%7D%7D%2C%7B%22component%22%3A%22status%22%2C%22x%22%3A0%2C%22y%22%3A5%2C%22w%22%3A2%2C%22h%22%3A1%7D%2C%7B%22component%22%3A%22spatial%22%2C%22coordinationScopes%22%3A%7B%22spatialZoom%22%3A%22A%22%2C%22spatialTargetX%22%3A%22A%22%2C%22spatialTargetY%22%3A%22A%22%2C%22spatialLayers%22%3A%22A%22%7D%2C%22x%22%3A2%2C%22y%22%3A0%2C%22w%22%3A5%2C%22h%22%3A8%7D%2C%7B%22component%22%3A%22spatial%22%2C%22coordinationScopes%22%3A%7B%22spatialZoom%22%3A%22B%22%2C%22spatialTargetX%22%3A%22A%22%2C%22spatialTargetY%22%3A%22A%22%2C%22spatialLayers%22%3A%22A%22%7D%2C%22x%22%3A7%2C%22y%22%3A0%2C%22w%22%3A5%2C%22h%22%3A8%7D%5D%7D)
- Pan-linked + zoom-offset-linked views of single image (linking of zoom delta rather than absolute zoom values)
    - Not currently possible within this PR. Would be possible with a "zoom offset" prop for the spatial component. Probably more appropriate for a separate issue/PR.

ilan-gold commented on src/components/heatmap/Heatmap.js:
```diff
@@ -358,7 +364,7 @@ const Heatmap = forwardRef((props, deckRef) => {
   const textLayers = [
     new HeatmapCompositeTextLayer({
-      showLoadingIndicator: backlog.length,
+      showLoadingIndicator: false,
```
What is happening here?

keller-mark replied:
I switched from showing the text "Loading..." as a heatmap DeckGL TextLayer to using a callback to be able to use the spinning animation from the parent TitleInfo component. So I can actually delete the `showLoadingIndicator` code entirely.

ilan-gold commented on src/components/heatmap/HeatmapTooltipSubscriber.js:
```diff
@@ -19,60 +19,56 @@ export default function HeatmapTooltipSubscriber(props) {
-        }
-      },
-    );
     const viewInfoToken = PubSub.subscribe(
```
I'm not sure what the solution here is, but this feels out of place given how clean everything else is.  Maybe in the docs or something we should note that tooltip events are somewhat outside of the scope of global state?

keller-mark replied:
I have cleaned up the tooltips a little bit. They still feel a little odd but I can make an issue.

ilan-gold commented on src/components/spatial/utils.js:
```diff
@@ -1,5 +1,215 @@
+  return result;
+}
+
+export async function initializeLayerChannelsIfMissing(layerDefsOrig, loaders) {
```
Can we have some comments here (and where this is used) about why this is needed?  For example, the reference on 135 to `channels` might seem odd if you didn't know that the purpose of it is to filter out other, non-raster layers (if I have that right).

keller-mark replied:
Done. This function allows the user to pass an array of channel selections which may be missing the `.slider`, `.domain`, `.color`, or `.visible` properties (the function will auto-initialize these properties if they are missing for any of the selected channels).

keller-mark commented on src/app/VitessceGrid.js:
```diff
@@ -123,19 +128,24 @@ export default function PubSubVitessceGrid(props) {
-        layout={config.staticLayout}
+      <VitessceGridLayout
+        layout={config.layout}
+        loaders={loaders}
```
This is a good point. I have removed the React context previously used to store the loaders instances and instead now store them in the `useStore` store.

keller-mark commented on src/app/DatasetLoaderProvider.js:
```diff
@@ -0,0 +1,62 @@
+    dataset.files.forEach((file) => {
+      const matchingLoaderClass = fileTypeToLoader[file.fileType];
+      if (!matchingLoaderClass) {
+        datasetLoaders.loaders[file.type] = new LoaderNotFoundError(
```
Right, I don't think this code makes much sense. I will update it.

keller-mark replied:
Instead of an error I have added a fallback to JsonLoader

keller-mark commented on src/app/view-config-utils.js:
```diff
@@ -0,0 +1,334 @@
+ * @param {string[]} prevScopes Previous scope names.
+ * @returns {string} The new scope name.
+ */
+export function getNextScope(prevScopes) {
```
Right they just need to be unique within the coordination object. So really they could be `String(Math.random())` or `uuidv4()` or something. I just thought it may be good to make them more human-readable and memorable since eventually we will want to expose a UI to update the coordination.

ilan-gold commented:
Testing this in the portal @keller-mark and it seems the z-slider is broken.  Do you want me to take a look at this within this PR or do you think you know what might be happening from the new code?

ilan-gold commented:
Another thing that might be unrelated, but might be related: 
![ezgif com-video-to-gif(6)](https://user-images.githubusercontent.com/43999641/91733979-50a7d280-eb78-11ea-81d4-3a7eea51bb68.gif)

If you don't know why any of these is happening off-hand, let me know and I can look into these things i find.

ilan-gold commented on src/app/Vitessce.js:
```diff
@@ -0,0 +1,154 @@
+    );
+  }
+  // Check if this is a "legacy" view config.
+  let upgradedConfig = config;
+  if (config.version === '0.1.0') {
+    // Validate under the legacy schema first.
+    const validateLegacy = new Ajv().compile(legacyConfigSchema);
+    const validLegacy = validateLegacy(config);
+
+    if (!validLegacy) {
+      const failureReason = JSON.stringify(validateLegacy.errors, null, 2);
+      return (
+        <Warning
+          title="Config validation failed"
+          preformatted={failureReason}
+          theme={theme}
+        />
+      );
+    }
+    // Upgrade from v0.1.0 to v1.0.0 before v1.0.0 schema validation.
+    upgradedConfig = upgrade(config);
+  }
```
I wonder if this is the issue with https://github.com/hubmapconsortium/vitessce/pull/728#issuecomment-683830818 since `upgrade` creates a new config object

ilan-gold replied:
One option could be to cache view states, but I don't know what that would look like -  hashing a string representation of the object?

keller-mark replied:
Oh I see, so the state is cleared every time the `<Vitessce/>` component re-renders, which may occur if the `height` prop changes (when going to full screen mode).

I think a solution would be to store the original view config prop in state or a ref and then diff against it during each render to determine whether it has updated. `useMemo` might work for this actually

ilan-gold replied:
Good point - we don't need to be aware of the history of past configs seen.  `useMemo` seems like a good solution (or something similar like you mentioned)

ilan-gold commented on src/components/layer-controller/RasterLayerController.js:
```diff
@@ -199,25 +150,20 @@ export default function RasterLayerController({
-      },
-    });
+    const newChannels = channels.map(c => ({ ...c, ...update }));
+    if (mouseUp) {
+      setChannels(newChannels);
+    }
```
I think we need something more advanced here (separate from the not-working aspect) - the controller doesn't work like it used to.  We can chat offline about this or I can take this on.  Basically, there needs to be some local state so the UI can update but then the actual layer only updates on `mouseUp`

ilan-gold commented on src/components/spatial/SpatialSubscriber.js:
```diff
@@ -2,210 +2,152 @@ import React, {
+  );
+
+  // Try to set up the layers array automatically if null or undefined.
+  useEffect(() => {
+    if (isReady && !layers) {
+      setLayers(sortLayers(autoLayers));
+    } else if (isReady && layers) {
+      // Layers were defined, but check whether channels for each layer were also defined.
+      // If channel / slider / domain definitions are missing, initialize in automatically.
+      initializeLayerChannelsIfMissing(layers, imageLayerLoaders)
+        .then(([newLayers, didInitialize]) => {
+          if (didInitialize) {
+            // Channels were only partially defined.
+            setLayers(newLayers);
+          }
+        });
+    }
+  }, [autoLayers, imageLayerLoaders, isReady, layers, setLayers]);
```
Why does this need to happen?  How could `useRasterData` return incomplete data, or is that not what this is for?

keller-mark replied:
The `else if` part is used in order to auto-initialize the `.slider`, `.domain`, `.visible`, and/or `.color` properties of pre-defined layer and channel definitions. This way, a user can specify the channel selections they want to view in the view config, but they can still get auto-initialization for sliders, etc.

For example, I used this feature with the Vanderbilt IMS image pyramid aggregation algorithm comparison view config. I just put the layer and channel selections in the view config and let the auto initialization do the rest.

```js
"spatialLayers": {
      "ORIGINAL": [
        {
          "type": "raster",
          "index": 0,
          "opacity": 1,
          "colormap": "",
          "channels": [
            { "selection": { "channel": 0, "z": 0, "time": 0 } }, // notice the missing "sliders" "domain" and "color" properties
            { "selection": { "channel": 1, "z": 0, "time": 0 } },
          ]
        }
      ],
```

however if I know those properties ahead of time I can specify (some or all of) them and those properties will not be overridden

```js
"spatialLayers": {
      "ORIGINAL": [
        {
          "type": "raster",
          "index": 0,
          "opacity": 1,
          "colormap": "",
          "channels": [
            { "selection": { "channel": 0, "z": 0, "time": 0 }, "slider": [0, 100], "domain": [0, 200] },
            { "selection": { "channel": 1, "z": 0, "time": 0 }, "slider": [0, 100], "domain": [0, 200], "visible": false, "color": [255, 0, 0] },
          ]
        }
      ],
```

ilan-gold replied:
Very cool!

ilan-gold commented:
@keller-mark As another architectural comment, can we store `viewInfo` that is currently being PubSub'ed in the global state as well?  I think it should be possible, and I can make a PR/trying it out for it into this branch if you think it makes sense.  It would allow us to remove a dependency, I think.

keller-mark commented on src/app/view-config-utils.js:
```diff
@@ -0,0 +1,334 @@
+
+  // The following coordination types should be
+  // initialized to independent scopes when
+  // initialized automatically.
```
I have moved this into `src/app/state/coordination.js` and added more comments there

keller-mark commented on src/app/state/hooks.js:
```diff
@@ -0,0 +1,70 @@
+  viewConfig: null,
+  setViewConfig: viewConfig => set({ viewConfig }),
+  setCoordinationValue: ({ parameter, scope, value }) => set(state => ({
+    viewConfig: {
```
I have added some comments here

keller-mark commented:
I just fixed a minor bug related to updating the scatterplot in response to cell radius computation. I tested in the portal and the view config state reset bug is fixed now. I made a branch here to add the `version` property to all view configs, which will need to be incorporated at the same time that vitessce is upgraded there https://github.com/hubmapconsortium/portal-ui/tree/keller-mark/vitessce-add-view-config-version

keller-mark commented:
I think I have responded to all of your feedback above @ilan-gold, except the removal of the VIEW_INFO event. I do agree that it would be great to remove the extra PubSub dependency. It is just that the tooltip implementation currently works like

```js
<ScatterplotSubscriber>
    <Scatterplot updateViewInfo={updateViewInfo}/>
    <ScatterplotTooltipSubscriber/>
</ScatterplotSubscriber>
```

where `updateViewInfo` is currently a call to PubSub.publish()

To convert this to use zustand instead of PubSub, I tried

```js
const useStore = create(set => ({
  viewInfo: null,
  updateViewInfo: (viewInfo, uuid) => set(state => ({
    viewInfo: { ...state.viewInfo, [uuid]: viewInfo }
  })),
}));
```

then `ScatterplotSubscriber` can consume the `updateViewInfo` setter function like

```js
const updateViewInfo = useStore(state => state.updateViewInfo);
```

while `ScatterplotTooltipSubscriber` can consume the `viewInfo` state like

```js
const viewInfo = useStore(state => state.viewInfo[uuid]);
```

However this causes `ScatterplotSubscriber` to do a unnecessary re-render every time `updateViewInfo` is called. This is because both the setter function and the value which it sets are _in_ the store. All consumers of the `useStore` hook need to re-render when the store changes, since the `useStore` return values may have changed.

But we really only want the consumer of the `viewInfo` _value_ to re-render (which in this case is the `ScatterplotTooltipSubscriber`). We know that running `updateViewInfo` will never change the `updateViewInfo` function, but react does not know that, since it is just another member of the store object. This is where the PubSub is helpful, since `PubSub.publish` is completely independent of `PubSub.subscribe` (when `ScatterplotTooltipSubscriber` is a subscriber, only `ScatterplotTooltipSubscriber` needs to re-render on each publish)


This zustand issue can be resolved if `ScatterplotTooltipSubscriber` is not a child of `ScatterplotSubscriber`, which we should do anyway (due to #735 / the z-index issues). Are you alright with saving the PubSub removal for another issue?

keller-mark commented:
Actually I am thinking about this more, and maybe it is because the `updateViewInfo` function _depends on_ the previous state parameter, so its function signature actually would need to change every time `viewInfo` changes... I wonder if there is an easy way to get around using the previous state in that setter function...

keller-mark commented:
Or perhaps I just need to read documentation more closely, it looks like there is a recommended solution for exactly this issue https://github.com/react-spring/zustand#transient-updates-for-often-occuring-state-changes

ilan-gold commented:
No worries @keller-mark - I'll test this out in the portal with all the changes while you dig into that a bit if you are going to.

ilan-gold commented:
@keller-mark I'm still noticing some pretty odd behavior in the channel controller so I'll need to make another PR for that.

keller-mark commented:
Ok, thanks for noticing this!

keller-mark commented:
I have removed the `pubsub-js` dependency. Going to test out in the portal and then if it looks good I will share an updated demo URL on slack.

keller-mark commented:
Updated demo: https://s3.amazonaws.com/vitessce-data/demos/2020-09-03/617abcc/index.html

ilan-gold commented:
These look good to me.  I made a PR to update the testing protocol which is a bit more general than this PR but is probably still a good thing to have.

mccalluc commented on NEXT.md:
```diff
@@ -0,0 +1,112 @@
+
+The implementation of coordinated multiple views here follows the coordination object model proposed by Boukhelifa and Rodgers (Information Visualization 2003).
+
+Below left is Figure 1 from Boukhelifa and Rodgers. Below right is that figure adapted to the Vitessce view config. ([view the diagram on google drawings here](https://docs.google.com/drawings/d/1jsNd2aG3OFlHfNzI3nfOl6UpMACw9JKyexCQUEd31fc/edit)).
+
+![Coordination object model in vitessce](https://user-images.githubusercontent.com/7525285/89790691-49008b00-daf0-11ea-95b5-cd74fe3499af.png)
```
```suggestion
Below left is Figure 1 from Boukhelifa and Rodgers. Below right is that figure adapted to the Vitessce view config. 

[![Coordination object model in vitessce](https://user-images.githubusercontent.com/7525285/89790691-49008b00-daf0-11ea-95b5-cd74fe3499af.png)](https://docs.google.com/drawings/d/1jsNd2aG3OFlHfNzI3nfOl6UpMACw9JKyexCQUEd31fc/edit)
```
So clicking on it takes you to the google doc... but you might feel this is less clear / more confusing. Either way.

mccalluc replied:
About the diagram: Does "A" in each of the coordinations refer to the same thing? I think that's not the case... each is its own namespace. 
- Perhaps use IDs with semantics for the example?
- Perhaps add lines to the diagram to make it clear what each "A" (or other ID) links?

mccalluc replied:
Diagram nitpick: I think using the same corner radius on all the round-rects will look tidier.

mccalluc commented on src/app/state/coordination.js:
```diff
@@ -0,0 +1,198 @@
+  EMBEDDING_TARGET_X: 'embeddingTargetX',
+  EMBEDDING_TARGET_Y: 'embeddingTargetY',
+  EMBEDDING_TARGET_Z: 'embeddingTargetZ',
+  SPATIAL_ZOOM: 'spatialZoom',
```
Can this support fixed zoom offsets? ie, view A is always 4x the zoom of view B?

keller-mark replied:
This has not been implemented in this current PR. But in theory, yes, this can support fixed zoom offsets, either a multiple like x4 or an added factor like +4.

I think the easiest approach to implement this would be to link view A and view B to the same zoom level, and then add two new `props` for the `spatial` component called `zoomOffset` and `zoomMultiple`. So both view A and view B would obtain the same zoom value from the coordination space, but view A would always multiply the value by 4 before passing into DeckGL (assuming it has the prop `zoomMultiple = 4`).

mccalluc commented on NEXT.md:
```diff
@@ -0,0 +1,112 @@
+
+### Coordination object graph representation
+
+One may think of the coordination objects, components, and scopes as forming a network or graph. It may be helpful to display this graph to the user.
```
This would be a bi-partite graph, with coordinations linked only to views, and views linked only to coordinations? Matrix representation, with coordinations on one axis and views on the other, might work.

keller-mark replied:
Right, so for a single coordination type the graph would look like this:

![Coordination model as graph representation - single coordination type](https://user-images.githubusercontent.com/7525285/92250949-1ef57b00-ee9a-11ea-80e3-9b5b668de38e.png)

<hr>
Considering more than one coordination type at a time may start to get messy. In the figure below, I have encoded the coordination type with a shape. Alternatively, you could imagine an additional column in the graph (tri-partite?) which links coordination types to coordination scopes. (For the below example, that column would contain two nodes, one for "rotation" and another for "zoom")

![Coordination model as graph representation - multiple coordination types](https://user-images.githubusercontent.com/7525285/92250958-2157d500-ee9a-11ea-90b3-f5746adef911.png)

keller-mark replied:
That extra column / tri-partite representation would look like
![Coordination model as graph representation - multiple coordination types - node encoding](https://user-images.githubusercontent.com/7525285/92252149-d6d75800-ee9b-11ea-8f82-11936e4f5076.png)

mccalluc reviewed (commented):
I've only looked at this really superficially, but I think it all looks good.

keller-mark commented on src/app/Vitessce.js:
```diff
@@ -0,0 +1,138 @@
+      if (!validLegacy) {
+        const failureReason = JSON.stringify(validateLegacy.errors, null, 2);
+        return [{
+          title: 'Config validation failed',
```
At a minimum version numbers will need to be added, yes. I have made those changes on a branch https://github.com/hubmapconsortium/portal-ui/tree/keller-mark/vitessce-add-view-config-version If we want to add linked view functionality in the portal then a more involved migration is required.

keller-mark commented on NEXT.md:
```diff
@@ -0,0 +1,112 @@
+This branch contains the code for a "next version" of Vitessce which will have the following key features:
```
Yes that is a good idea. I will try to clean the notes up a bit when moving them into README. And yes I will think about how to update that figure. I think conceptually zustand can be thought about like pubsub but there are probably minor changes to the figure that would make it more accurate.

ilan-gold commented:
@keller-mark I am wondering what the status of this is?  Do we need to have a chat to clarify anything?

In addition, I just realized something - do we have the option to "save state?"  I am not sure what this would look like - perhaps a hook into the `zustand` store that allows you to hoist the the state up a level (or many levels) for exporting from the portal?

keller-mark commented:
Did you want to merge #745 first?

For saving state, the outer application can listen for the `onConfigChange` via the callback on the Vitessce component https://github.com/hubmapconsortium/vitessce/blob/next/src/app/Vitessce.js#L46

ilan-gold commented:
I will merge now yes @keller-mark and thank you for that pointer.

keller-mark commented on src/schemas/config.schema.json:
```diff
@@ -0,0 +1,336 @@
+        }
+      }
+    },
+    "requestInit": {
```
Thanks, I have noted this in the schema description field.

keller-mark commented:
I have removed the `NEXT.md` file and updated the `GLOSSARY.md` with definitions of coordination-related terms https://github.com/hubmapconsortium/vitessce/blob/next/GLOSSARY.md
